File: src/main/java/com/bibireden/data_attributes/impl/EntityAttributeData.java
package com.bibireden.data_attributes.impl;

import java.io.IOException;
import java.util.HashMap;
import java.util.Map;
import java.util.function.BiFunction;

import com.bibireden.data_attributes.data.AttributeOverride;
import com.bibireden.data_attributes.endec.NbtDeserializer;
import com.bibireden.data_attributes.endec.NbtSerializer;
import com.bibireden.data_attributes.json.AttributeFunctionJson;
import com.bibireden.data_attributes.mutable.MutableEntityAttribute;

import net.minecraft.entity.attribute.EntityAttribute;
import net.minecraft.nbt.NbtCompound;
import net.minecraft.util.Identifier;
import net.minecraft.registry.Registries;

public final class EntityAttributeData implements NbtIO {
	private AttributeOverride attribute;
	private final Map<Identifier, AttributeFunctionJson> functions;

	public EntityAttributeData() {
		this.functions = new HashMap<>();
	}

	public EntityAttributeData(final AttributeOverride attribute) {
		this();
		this.attribute = attribute;
	}

	public void override(final Identifier identifier, BiFunction<Identifier, EntityAttribute, EntityAttribute> function) {
		if (this.attribute == null) return;
		EntityAttribute entityAttribute = function.apply(identifier, this.attribute.create());
		this.attribute.override((MutableEntityAttribute) entityAttribute);
	}

	public void copy(EntityAttribute entityAttributeIn) {
		MutableEntityAttribute mutableEntityAttribute = (MutableEntityAttribute) entityAttributeIn;

		for (Identifier identifier : this.functions.keySet()) {
			EntityAttribute entityAttribute = Registries.ATTRIBUTE.get(identifier);

			if (entityAttribute == null) continue;

			AttributeFunctionJson function = this.functions.get(identifier);
			mutableEntityAttribute.addChild((MutableEntityAttribute) entityAttribute, function);
		}
	}

	public void putFunctions(Map<Identifier, AttributeFunctionJson> functions) {
		this.functions.putAll(functions);
	}

	@Override
	public void readFromNbt(NbtCompound tag) throws IllegalStateException {
		if (tag.contains("Attribute")) {
			this.attribute = AttributeOverride.Companion.getEndec().decodeFully(NbtDeserializer::of, tag.getCompound("Attribute"));
        }

		NbtCompound functions = tag.getCompound("Functions");
		functions.getKeys().forEach(key -> this.functions.put(new Identifier(key), AttributeFunctionJson.read(functions.getByteArray(key))));
	}

	@Override
	public void writeToNbt(NbtCompound tag) {
		if (this.attribute != null) {
			tag.put("Attribute", AttributeOverride.Companion.getEndec().encodeFully(NbtSerializer::of, this.attribute));
		}

		NbtCompound functions = new NbtCompound();
		this.functions.forEach((key, value) -> functions.putByteArray(key.toString(), value.write()));
		tag.put("Functions", functions);
	}
}

File: src/main/kotlin/com/bibireden/data_attributes/data/EntityAttributeData.kt
package com.bibireden.data_attributes.data

import com.bibireden.data_attributes.DataAttributes
import com.bibireden.data_attributes.config.functions.AttributeFunction
import com.bibireden.data_attributes.config.models.OverridesConfigModel.AttributeOverride
import com.bibireden.data_attributes.endec.Endecs
import com.bibireden.data_attributes.ext.keyOf
import com.bibireden.data_attributes.mutable.MutableEntityAttribute
import io.wispforest.endec.impl.StructEndecBuilder
import net.minecraft.entity.attribute.EntityAttribute
import net.minecraft.registry.Registries
import net.minecraft.util.Identifier

/**
 * Data that might contain a `AttributeOverrideConfig` and contains
 * `AttributeFunctions` to be used to override or copy to `EntityAttributes`.
 */
class EntityAttributeData(val override: AttributeOverride? = null, val functions: MutableMap<Identifier, AttributeFunction> = mutableMapOf()) {
    constructor(value: AttributeOverride) : this(value, mutableMapOf())

    companion object {
        @JvmField
        val ENDEC = StructEndecBuilder.of(
            AttributeOverride.ENDEC.nullableOf().fieldOf("override") { it.override },
            Endecs.IDENTIFIER.keyOf(AttributeFunction.ENDEC).fieldOf("functions") { it.functions },
            ::EntityAttributeData,
        )
    }

    /**
     * Sets the override for an [EntityAttribute].
     *
     * This will completely overwrite any other override set and replace it with the one present in this class.
     * If there is none present, then no override will be done.
     */
    fun override(attribute: EntityAttribute) {
        this.override?.override(attribute)
    }

    /** Copies to a provided [EntityAttribute] by adding children to itself. */
    fun copy(attribute: EntityAttribute) {
        with(attribute as MutableEntityAttribute) {
            for ((id, function) in functions) {
                val childAttribute = Registries.ATTRIBUTE[id] as? MutableEntityAttribute ?: continue
                this.`data_attributes$addChild`(childAttribute, function)
            }
        }
    }
    
    /** Joins a [List] of [AttributeFunction]'s with the data in this class. */
    fun putFunctions(functions: List<AttributeFunction>) {
        this.functions.putAll(functions.map { (id, behavior, value) ->
            if (!Registries.ATTRIBUTE.containsId(id)) {
                DataAttributes.LOGGER.warn("The attribute function child [$id] does not seem to be registered. This could allude to a missing mod or registered attribute.")
            }
            id to AttributeFunction(id, behavior, value)
        })
    }
}